### ISSUE TITLE
Getting Runtime error instead of compilation error

### ISSUE THREAD

**Issue by soninishank** [mapstruct/mapstruct#2661]
java: Internal error in the mapping processor: java.lang.RuntimeException: org.mapstruct.ap.shaded.freemarker.core.InvalidReferenceException: The following has evaluated to null or missing:  ==> ext.targetBeanName  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 31, column 12]    ----  Tip: It's the step after the last dot that caused this error, not those before it.  ----  Tip: If the failing expression is known to be legally refer to something that's null or missing, either specify a default value like myOptionalVar!myDefault, or use <#if myOptionalVar??>when-present<#else>when-missing</#if>. (These only cover the last step of the expression; to cover the whole expression, use parenthesis: (myOptionalVar.foo)!myDefault, (myOptionalVar.foo)??  ----    ----  FTL stack trace ("~" means nesting-related):  	- Failed at: ${ext.targetBeanName}  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" in macro "assignToExistingTarget" at line 31, column 10]  	- Reached through: @assignToExistingTarget  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 14, column 7]  	~ Reached through: #nested  [in template "org/mapstruct/ap/internal/model/macro/CommonMacros.ftl" in macro "handleExceptions" at line 94, column 9]  	~ Reached through: @lib.handleExceptions  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 10, column 1]  ----  	at org.mapstruct.ap.internal.writer.ModelWriter.writeModel(ModelWriter.java:77)  	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.createSourceFile(MapperRenderingProcessor.java:62)  	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.writeToSourceFile(MapperRenderingProcessor.java:39)  	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.process(MapperRenderingProcessor.java:29)  	at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.process(MapperRenderingProcessor.java:24)  	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:338)  	at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:318)  	at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:267)  	at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:166)  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:985)  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:901)  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1227)  	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1340)  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1258)  	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:936)  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)  	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)  	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:238)  	at org.jetbrains.jps.incremental.java.JavaBuilder.lambda$compileJava$2(JavaBuilder.java:514)  	at org.jetbrains.jps.incremental.java.JavaBuilder.invokeJavac(JavaBuilder.java:560)  	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:512)  	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:355)  	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:280)  	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:234)  	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1485)  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1123)  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1268)  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:1088)  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:854)  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:441)  	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:190)  	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:132)  	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:318)  	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:146)  	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:218)  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)  	at java.base/java.lang.Thread.run(Thread.java:829)  Caused by: org.mapstruct.ap.shaded.freemarker.core.InvalidReferenceException: The following has evaluated to null or missing:  ==> ext.targetBeanName  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 31, column 12]    ----  Tip: It's the step after the last dot that caused this error, not those before it.  ----  Tip: If the failing expression is known to be legally refer to something that's null or missing, either specify a default value like myOptionalVar!myDefault, or use <#if myOptionalVar??>when-present<#else>when-missing</#if>. (These only cover the last step of the expression; to cover the whole expression, use parenthesis: (myOptionalVar.foo)!myDefault, (myOptionalVar.foo)??  ----    ----  FTL stack trace ("~" means nesting-related):  	- Failed at: ${ext.targetBeanName}  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" in macro "assignToExistingTarget" at line 31, column 10]  	- Reached through: @assignToExistingTarget  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 14, column 7]  	~ Reached through: #nested  [in template "org/mapstruct/ap/internal/model/macro/CommonMacros.ftl" in macro "handleExceptions" at line 94, column 9]  	~ Reached through: @lib.handleExceptions  [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 10, column 1]  ----  	at org.mapstruct.ap.shaded.freemarker.core.InvalidReferenceException.getInstance(InvalidReferenceException.java:108)  	at org.mapstruct.ap.shaded.freemarker.core.EvalUtil.coerceModelToString(EvalUtil.java:346)  	at org.mapstruct.ap.shaded.freemarker.core.Expression.evalAndCoerceToString(Expression.java:80)  	at org.mapstruct.ap.shaded.freemarker.core.DollarVariable.accept(DollarVariable.java:40)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Macro$Context.runMacro(Macro.java:173)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:686)  	at org.mapstruct.ap.shaded.freemarker.core.UnifiedCall.accept(UnifiedCall.java:80)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.IfBlock.accept(IfBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:479)  	at org.mapstruct.ap.shaded.freemarker.core.BodyInstruction.accept(BodyInstruction.java:56)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.IfBlock.accept(IfBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Macro$Context.runMacro(Macro.java:173)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:686)  	at org.mapstruct.ap.shaded.freemarker.core.UnifiedCall.accept(UnifiedCall.java:80)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.process(Environment.java:235)  	at org.mapstruct.ap.shaded.freemarker.template.Template.process(Template.java:262)  	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter.write(FreeMarkerModelElementWriter.java:35)  	at org.mapstruct.ap.internal.writer.FreeMarkerWritable.write(FreeMarkerWritable.java:19)  	at org.mapstruct.ap.internal.writer.ModelIncludeDirective.execute(ModelIncludeDirective.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:333)  	at org.mapstruct.ap.shaded.freemarker.core.UnifiedCall.accept(UnifiedCall.java:100)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.process(Environment.java:235)  	at org.mapstruct.ap.shaded.freemarker.template.Template.process(Template.java:262)  	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter.write(FreeMarkerModelElementWriter.java:35)  	at org.mapstruct.ap.internal.writer.FreeMarkerWritable.write(FreeMarkerWritable.java:19)  	at org.mapstruct.ap.internal.writer.ModelIncludeDirective.execute(ModelIncludeDirective.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:333)  	at org.mapstruct.ap.shaded.freemarker.core.UnifiedCall.accept(UnifiedCall.java:100)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock$Context.runLoop(IteratorBlock.java:145)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitIteratorBlock(Environment.java:501)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock.accept(IteratorBlock.java:67)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.ConditionalBlock.accept(ConditionalBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock$Context.runLoop(IteratorBlock.java:145)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitIteratorBlock(Environment.java:501)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock.accept(IteratorBlock.java:67)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.IfBlock.accept(IfBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.IfBlock.accept(IfBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitByHiddingParent(Environment.java:278)  	at org.mapstruct.ap.shaded.freemarker.core.ConditionalBlock.accept(ConditionalBlock.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.process(Environment.java:235)  	at org.mapstruct.ap.shaded.freemarker.template.Template.process(Template.java:262)  	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter.write(FreeMarkerModelElementWriter.java:35)  	at org.mapstruct.ap.internal.writer.FreeMarkerWritable.write(FreeMarkerWritable.java:19)  	at org.mapstruct.ap.internal.writer.ModelIncludeDirective.execute(ModelIncludeDirective.java:48)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:333)  	at org.mapstruct.ap.shaded.freemarker.core.UnifiedCall.accept(UnifiedCall.java:100)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock$Context.runLoop(IteratorBlock.java:145)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visitIteratorBlock(Environment.java:501)  	at org.mapstruct.ap.shaded.freemarker.core.IteratorBlock.accept(IteratorBlock.java:67)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.MixedContent.accept(MixedContent.java:57)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.visit(Environment.java:257)  	at org.mapstruct.ap.shaded.freemarker.core.Environment.process(Environment.java:235)  	at org.mapstruct.ap.shaded.freemarker.template.Template.process(Template.java:262)  	at org.mapstruct.ap.internal.writer.FreeMarkerModelElementWriter.write(FreeMarkerModelElementWriter.java:35)  	at org.mapstruct.ap.internal.writer.FreeMarkerWritable.write(FreeMarkerWritable.java:19)  	at org.mapstruct.ap.internal.writer.ModelWriter.writeModel(ModelWriter.java:69)  	... 39 more

**Comment by filiphr** [comment]
Thanks for raising this issue @soninishank. Can you please share your Mapper that causes this problem and also the version of MapStruct that you are using?

**Comment by soninishank** [comment]
<mapstruct.version>1.4.1.Final</mapstruct.version>

**Comment by soninishank** [comment]
Actually, this is my office code, I am not sure how can I share that with you.

**Comment by soninishank** [comment]
It is pointing me to this :

@Mapper(componentModel = "spring", nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS, unmappedTargetPolicy = ReportingPolicy.IGNORE)
public **interface** ReleaseFundDetailsMapper

**Comment by Zegveld** [comment]
Can you check if this crash still happens with `1.5.0.Beta1`?

In case it does still happen we need more information than just the interface definition to analyze this. One advantage is that you have already pinpointed which Mapper is the cause of the crash, now we still need to figure out which mapping and with which annotations.

If you could create a reproduction scenario where you have the same situation, but without information that would be related to your work would help a lot. One way of doing this is for example renaming all the Source/Target classes to be called Source#/Target# and all the fields to be called field#. (where matching field names have the same name in source/target)

If you could also downsize (remove all irrelevant fields/classes to the crash) the reproduction scenario so that we get it as small as possible, that would be great. This will help us with analyzing the crash, because it will limit the times we would pass a specific bit of code during debugging. Debugging a mapper that maps 5 fields is quicker then one that maps 100 fields. If it is not downsized that is also fine.

I've tried to create a small reproduction with the information available but did not succeed at it. What I figured out by looking at the stack-trace is that it has something to do with exception handling. Apparently when the exception is thrown somehow it doesn't know for which field it did the check.

cleaned up a bit of the stack-trace for readability:
```
The following has evaluated to null or missing: ==> ext.targetBeanName [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 31, column 12] 
---- FTL stack trace ("" means nesting-related): 
- Failed at: ${ext.targetBeanName} [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" in macro "assignToExistingTarget" at line 31, column 10] 
- Reached through: @assignToExistingTarget [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 14, column 7] 
~ Reached through: #nested [in template "org/mapstruct/ap/internal/model/macro/CommonMacros.ftl" in macro "handleExceptions" at line 94, column 9] 
~ Reached through: @lib.handleExceptions [in template "org/mapstruct/ap/internal/model/assignment/UpdateWrapper.ftl" at line 10, column 1] 
```

I hope the above information (not the stack trace) helps with creating a reproduction scenario, because without one we do not know if we have fixed it nor what exactly needs fixing.

**Comment by tobiascervin** [comment]
I'm got the exact same error just now using:
<org.mapstruct.version>1.4.2.Final</org.mapstruct.version>
<org.projectlombok.version>1.18.18</org.projectlombok.version>

Still getting the same error when updating to 1.5.0.Beta1

**Comment by tobiascervin** [comment]
`class Event 
{
  private String senderName
  private String receiverName
  private PartyDetails senderDetails
  private PartyDetails receiverDetails
}

@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
  public interface EventMapper {
    EventMapper MAPPER = Mappers.getMapper(
            EventMapper.class
    );
    @Mapping(target = "receiver.name", source = "receiverName")
    @Mapping(target = "receiver.alias", source = "receiverDetails.alias")
    EventRepresentation map(Event event);
}
`

It works if I remove the mapping done from the *Details property.

Also now I tried to take away Lomboks @Builder (Only have @Data now) annotation from the "EventRepresentation" and then it built Successfully!

**Comment by Zegveld** [comment]
Thanks for the reproduction scenario @tobiascervin. 
I'm going to attempt to reproduce it locally.

ps. I've updated your comment so that the code blocks are correctly marked.

**Comment by tobiascervin** [comment]
Thanks, first time contributing here :D.

**Comment by Zegveld** [comment]
Ok I've finished with the analysis of the situation.

If there is a builder present the following will not work
```
    @Mapping( target = "receiver.name", source = "receiverName" )
    @Mapping( target = "receiver.alias", source = "receiverDetails.alias" )
    EventRepresentation map(Event event);
```

But when you split it into 2 mapping methods it does work:
```
    @Mapping( target = "receiver", source = "." )
    EventRepresentation mapEventToEventRepresentation(Event event);

    @Mapping( target = "name", source = "receiverName" )
    @Mapping( target = "alias", source = "receiverDetails.alias" )
    EventReceiver mapEventToEventReceiver(Event event);
```

This seems to be related in how mapstruct handles nested mappings.
Apparently during nested mappings mapstruct does not keep track of the targetBeanName correctly, or it somehow becomes lost.

I might look further into it later, or someone else can do that.
As a work-around for now: when you get this crash then split up the nested mappings as demonstrated above.

**Comment by filiphr** [comment]
Great analysis @Zegveld. This is a tricky problem, nested target mappings are quite tricky, especially splits like this.

**Comment by esbug** [comment]
Hello, has this problem been solved, I also encountered the same problem, and an error was reported when compiling.

mapstruc version: 1.5.2.Final

**Comment by vishalgupta-valtech** [comment]
I had this issue as well few days ago. I split the mappings into 2 mapping methods.  I used above example and it worked

  ```
@Mapping( target = "receiver", source = "." )
    EventRepresentation mapEventToEventRepresentation(Event event);
    @Mapping( target = "name", source = "receiverName" )
    @Mapping( target = "alias", source = "receiverDetails.alias" )
    EventReceiver mapEventToEventReceiver(Event event);
```

**Comment by thunderhook** [comment]
This might be solved with #2941 when nested target properties are handled differently.

**Comment by filiphr** [comment]
Thanks for tracking this down @thunderhook. I'll close this in favour of #2941
